### ISSUE TITLE
Installer: Improve dependency handling

### ISSUE THREAD

**Issue by Rondom** [dokan-dev/dokany#322]
Again I am not familiar with MSI, so take this feature request with a grain of salt.

Downstream users of Dokan currently install Dokan Library during installation of their software product. When they upgrade to a new major version Dokan (for example from Dokan 1.x.x to 2.y.y) they cannot safely remove the older version because it could break other applications.

This issue is for investigating in how far Windows Installer provides features to declare a dependency on Dokan and handle this cleanly (Components?, Merge Modules?) and implementing a possible solution.

**Comment by js69** [comment]
This is something that can comfortably only be handled by a custom bootstrapper. MSI itself offers no real mechanism to restart the installation after a reboot.

**Comment by Rondom** [comment]
There is no reboot needed when doing a major upgrade e.g. 1.x.x to 2.y.y. The new driver and library can be installed alongside the old one.

**Comment by js69** [comment]
Regarding uninstall of unneeded software: there is a dependency registration on MSI packages, but that registration is only really done by WiX bundles so you would not really know if there isn't other software that uses an older Dokan version.

In the WiX bundle install log you'll see this:
Registering dependency: {805152fc-59e3-4585-9f11-ea59b34f36b9} on package provider: {65A3A964-3DC3-0100-0000-160729112819}, package: Dokan_x64.msi

This creates an entry here:
HKEY_CLASSES_ROOT\Installer\Dependencies\{65A3A964-3DC3-0100-0000-160729112819}\Dependents\{805152fc-59e3-4585-9f11-ea59b34f36b9}

WRT to scenarios like MediaPortal:
You could detect an obsolete Dokan library in your application's bundle setup, but should only offer an option to uninstall, not force it.

**Comment by overheadhunter** [comment]
Maybe this is a matter of documentation in the [How to package your application with Dokan](https://github.com/dokan-dev/dokany/wiki/How-to-package-your-application-with-Dokan):

We could use some kind of matrix showing _what_ to package on what Windows version and CPU architecture.

For example: Afaik for the latest Dokany 1.3.0.1000 on Windows 10 64bit only one single VC++ redistributable is required. But I am not sure and would like some kind of official documentation for this, so I know what I have to check in our downstream installer.

**Comment by vrs** [comment]
In my experience the easiest and most robust approach for something that's only libraries (so a few dlls and resources) is to not bother with merge modules and instead simply vendor the files you need into your app's installation. That way you don't have to deal with MSI refcount, versioning semantics, and any kind of shared resource. You'll have dll search order issues if something was installed as a KnownDLL or into System32, but doing that is not really good practice anyway.

**Comment by LTRData** [comment]
> In my experience the easiest and most robust approach for something that's only libraries (so a few dlls and resources) is to not bother with merge modules and instead simply vendor the files you need into your app's installation. That way you don't have to deal with MSI refcount, versioning semantics, and any kind of shared resource. You'll have dll search order issues if something was installed as a KnownDLL or into System32, but doing that is not really good practice anyway.

Dokan is not only a few dlls and resources, it contains a driver. This means that it needs to be installed in a special way and copied into system32\drivers directory. It cannot just be installed into an application directory and referenced from there. So, since it contains shared components, care needs to be taken to not cause issues for other applications using the same major version of Dokan when installing, uninstalling and upgrading.

**Comment by vrs** [comment]
> Dokan is not only a few dlls and resources, it contains a driver. This means that it needs to be installed in a special way and copied into system32\drivers directory.
Yes. My recommendation is only for apps that don't install the driver or manage it out of band, which may be valid for some.

**Comment by Liryna** [comment]
For anyone reading this later, even if you install the driver from another location on the system that does not conflict with the official installer, you will still conflict with the driver resources. To avoid that, change the constant, build and sign your own driver https://github.com/dokan-dev/dokany/blob/0870b26969b505d53e8e63354a7be6343b2af0e7/sys/dokan.h#L63-L91
